server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java

```diff
@@ -192,9 +197,41 @@ ClusterState execute(ClusterState currentState) {
             }
 
             if (skippedSettings.isEmpty() == false && openIndices.isEmpty() == false) {
-                throw new IllegalArgumentException(
-                    String.format(Locale.ROOT, "Can't update non dynamic settings [%s] for open indices %s", skippedSettings, openIndices)
-                );
+                if (request.reopenShards()) {
+                    // We have non-dynamic settings and open indices. We will unassign all of the shards in these indices so that the new
+                    // changed settings are applied when the shards are re-assigned.
+                    routingTableBuilder = RoutingTable.builder(
+                        allocationService.getShardRoutingRoleStrategy(),
+                        currentState.routingTable()
+                    );
+                    for (Index index : openIndices) {
+                        List<ShardRouting> shardRoutingList = currentState.routingTable().allShards(index.getName());
+                        IndexRoutingTable.Builder indexRoutingTableBuilder = IndexRoutingTable.builder(index);
+                        for (ShardRouting shardRouting : shardRoutingList) {
+                            if (ShardRoutingState.UNASSIGNED.equals(shardRouting.state()) == false) {
+                                indexRoutingTableBuilder.addShard(
+                                    shardRouting.moveToUnassigned(
+                                        new UnassignedInfo(UnassignedInfo.Reason.INDEX_CLOSED, "Unassigning shards")
```

Comment:
Would `INDEX_REOPENED` be a more appropriate reason here?

Reply:
Oh I meant to reach out about this! Is INDEX_CLOSED appropriate? Or do we need a new reason like 'UPDATING_SETTINGS`?

Reply:
It doesn't really matter in this case AFAICT. We could make a new value but then we'd need a fallback for bwc (see `org.elasticsearch.cluster.routing.UnassignedInfo#writeTo` for similar cases). But since we're calling this feature "reopen" in the API, `INDEX_REOPENED` seems like a good choice.